rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/utils/JAXRSUtils.java

```diff
@@ -1508,11 +1509,40 @@ public static boolean matchMimeTypes(MediaType requestContentType,
                 }
             }
         }
-
         return new ArrayList<>(supportedMimeTypeList);
+    }
+
+    public static boolean doMimeTypesIntersect(List<MediaType> requiredMediaTypes, List<MediaType> userMediaTypes) {
+        for (MediaType requiredType : requiredMediaTypes) {
```

Comment:
Regarding this part, I understand the purpose, but now there 2 identical places, `intersectMimeTypes` and `doMimeTypesIntersect`, with the same code (solving one problem but introducing another). What I am thinking about is to introduce some kind of accumulator to either map the result of the operation to the `List` or `true/false`, do you think it would be possible @WhiteCat22 ?

Reply:
The performance increase comes from not making a HashSet just to do the comparison. I agree that duplicating code is not ideal, but I'm not sure how to accomplish both cases with a single method.

I am open to suggestions.

Reply:
Turned out to be quite tricky ... but I have one suggestion for you. Here is the implementation for `doMimeTypesIntersect` and `intersectMimeTypes`:

```
  public static List<MediaType> intersectMimeTypes(List<MediaType> requiredMediaTypes,
                                                     List<MediaType> userMediaTypes,
                                                     boolean addRequiredParamsIfPossible,
                                                     boolean addDistanceParameter) {
        final AccumulatingConsumer consumer = new AccumulatingConsumer(addRequiredParamsIfPossible, addDistanceParameter);
        intersectMimeTypes(requiredMediaTypes, userMediaTypes, consumer);
        return new ArrayList<>(consumer.supportedMimeTypeList);
    }
    
    public static boolean doMimeTypesIntersect(List<MediaType> requiredMediaTypes, List<MediaType> userMediaTypes) {
        final NonAccumulatingConsumer consumer = new NonAccumulatingConsumer();
        intersectMimeTypes(requiredMediaTypes, userMediaTypes, consumer);
        return consumer.doIntersect;
    }
```
Here is the initial `intersectMimeTypes` function, refactored to use consumers, the part which is was accumulating (or returning `true`) is now part of the consumers:

```
private static void intersectMimeTypes(List<MediaType> requiredMediaTypes, List<MediaType> userMediaTypes,
            BiConsumer<MediaType, MediaType> consumer) {

        for (MediaType requiredType : requiredMediaTypes) {
            for (MediaType userType : userMediaTypes) {
                boolean isCompatible = isMediaTypeCompatible(requiredType, userType);
                if (isCompatible) {
                    boolean parametersMatched = true;
                    for (Map.Entry<String, String> entry : userType.getParameters().entrySet()) {
                        String value = requiredType.getParameters().get(entry.getKey());
                        if (value != null && entry.getValue() != null
                            && !(stripDoubleQuotesIfNeeded(value).equals(
                                    stripDoubleQuotesIfNeeded(entry.getValue())))) {
                            
                            if (HTTP_CHARSET_PARAM.equals(entry.getKey())
                                && value.equalsIgnoreCase(entry.getValue())) {
                                continue;
                            }
                            parametersMatched = false;
                            break;
                        }
                    }
                    if (!parametersMatched) {
                        continue;
                    }
                    
                    consumer.accept(requiredType, userType);
                }
            }
        }
    }

```

And finally, two consumers:
```
    private static class NonAccumulatingConsumer implements BiConsumer<MediaType, MediaType> {
        private boolean doIntersect;
        
        @Override
        public void accept(MediaType t, MediaType u) {
            doIntersect = true;
        }
    }
```

And 
```
    private static class AccumulatingConsumer implements BiConsumer<MediaType, MediaType> {
        private final Set<MediaType> supportedMimeTypeList = new LinkedHashSet<MediaType>();
        private final boolean addRequiredParamsIfPossible;
        private final boolean addDistanceParameter;
        
        private AccumulatingConsumer(boolean addRequiredParamsIfPossible, boolean addDistanceParameter) {
            this.addRequiredParamsIfPossible = addRequiredParamsIfPossible;
            this.addDistanceParameter = addDistanceParameter;
        }
        
        @Override
        public void accept(MediaType requiredType, MediaType userType) {
            boolean requiredTypeWildcard = requiredType.getType().equals(MediaType.MEDIA_TYPE_WILDCARD);
            boolean requiredSubTypeWildcard = requiredType.getSubtype().contains(MediaType.MEDIA_TYPE_WILDCARD);

            String type = requiredTypeWildcard ? userType.getType() : requiredType.getType();
            String subtype = requiredSubTypeWildcard ? userType.getSubtype() : requiredType.getSubtype();

            Map<String, String> parameters = userType.getParameters();
            if (addRequiredParamsIfPossible) {
                parameters = new LinkedHashMap<String, String>(parameters);
                for (Map.Entry<String, String> entry : requiredType.getParameters().entrySet()) {
                    if (!parameters.containsKey(entry.getKey())) {
                        parameters.put(entry.getKey(), entry.getValue());
                    }
                }
            }
            if (addDistanceParameter) {
                int distance = 0;
                if (requiredTypeWildcard) {
                    distance++;
                }
                if (requiredSubTypeWildcard) {
                    distance++;
                }
                parameters.put(MEDIA_TYPE_DISTANCE_PARAM, Integer.toString(distance));
            }
            supportedMimeTypeList.add(new MediaType(type, subtype, parameters));
        }
    }
```

Ideally, would be good to move to a dedicated utility class, but roughly this is an idea. What do you think. @WhiteCat22, does it make sense?

Reply:
This makes sense, but is there another interface that we can use instead of BiConsumer? BiConsumer requires Java 1.8 and this is something that we would like to see backported to 3.1.X which has 1.7 dependencies. Another thing is that we need to have a mechanism to return after the first match during NonAccumulatingConsumer .accept() instead of going through the entire double for loops. There are several ways to do this, pass in a flag to allow us to bail out after first match, return a boolean from accept, or create a new isFinished() method to let us know that we are done processing.

Reply:
Yeah, sure, sounds good, something like that would fit:
```
public interface MimeTypesIntersector {
    boolean void intersect(MediaType requiredType, MediaType userType);
}
```

And than

```
private static void intersectMimeTypes(List<MediaType> requiredMediaTypes, List<MediaType> userMediaTypes,
MimeTypesIntersector intersector) {
    ... 
   if (!intersector.intersect( requiredType, userType) {
       return;
   }
}
```

Reply:
Alright, I've had a chance to code this up. Should these classes each go in their own class in the `org.apache.cxf.jaxrs.utils` package, or a single util class?

Reply:
I would vote for own class, what do you think?

Reply:
I agree, I have them coded up into separate `MimeTypesIntersector`, `AccumulatingIntersector` and `NonAccumulatingIntersector` classes

Reply:
:+1: